### PR TITLE
Schema v2 prevent shutdown

### PR DESCRIPTION
**Pull request by aliszka**
### What's being changed:
Adds locks to prevent concurrent shard usage (handling requests) and shutdown

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/repos/db/index.go
Comment (jeroiraz) [inline adapters/repos/db/index.go:647]: just wondering if `backupMutex` could be used to prevent shutdown as it's being already acquired during insertions...

File: adapters/repos/db/index.go
Comment (jeroiraz) [inline adapters/repos/db/index.go:677]: is was not in the original code but it may be worth returning the error returned by `getOrInitLocalShardNoShutdown`, it would require validating that `ErrShardNotFound` is actually returned

File: adapters/repos/db/index.go
Comment (jeroiraz) [inline adapters/repos/db/index.go:903]: same comment as before, probably it's better to return `err` to avoid hiding a different error that could be returned by `getOrInitLocalShardNoShutdown`

File: adapters/repos/db/index.go
Comment (jeroiraz) [inline adapters/repos/db/index.go:978]: also same comment here

File: adapters/repos/db/index.go
Comment (jeroiraz) [inline adapters/repos/db/index.go:1043]: same comment

File: adapters/repos/db/index.go
Comment (jeroiraz) [inline adapters/repos/db/index.go:1055]: same comment

File: adapters/repos/db/index.go
Comment (jeroiraz) [inline adapters/repos/db/index.go:1191]: same comment

File: adapters/repos/db/index.go
Comment (jeroiraz) [inline adapters/repos/db/index.go:1614]: same here

File: adapters/repos/db/index.go
Comment (jeroiraz) [inline adapters/repos/db/index.go:1690]: same here

File: adapters/repos/db/index.go
Comment (jeroiraz) [inline adapters/repos/db/index.go:1819]: same here

File: adapters/repos/db/index.go
Comment (jeroiraz) [inline adapters/repos/db/index.go:1872]: same here

File: adapters/repos/db/index.go
Comment (jeroiraz) [inline adapters/repos/db/index.go:2043]: same here

File: adapters/repos/db/index.go
Comment (jeroiraz) [inline adapters/repos/db/index.go:2098]: same here

File: adapters/repos/db/index.go
Comment (jeroiraz) [inline adapters/repos/db/index.go:2120]: same here

File: adapters/repos/db/shard.go
Comment (moogacs) [inline adapters/repos/db/shard.go:1071]: **thought:**  what do you think if we log if `ctx.Err()` != nil and always return nil, because if the ctx got cancelled for reasons like was finished etc. it's still not an error

File: adapters/repos/db/index.go
Comment (jeroiraz) [inline adapters/repos/db/index.go:2179]: same here

File: adapters/repos/db/index.go
Comment (jeroiraz) [inline adapters/repos/db/index.go:2255]: same here

File: adapters/repos/db/migrator.go
Comment (jeroiraz) [inline adapters/repos/db/migrator.go:355]: not 100% sure if it's the equivalent behaviour, as at the time `if shard == nil` is evaluated no lock is being acquired so the shard could have changed its state meantime

File: adapters/repos/db/index.go
Comment (aliszka) [reply adapters/repos/db/index.go:647]: mutex is acquired for the entire class/index being backed up, so its purpose is a bit different.
but I can agree, that we have to many locks at the moment and need to reduce that number soon as code is getting harder to maintain

File: adapters/repos/db/index.go
Comment (aliszka) [reply adapters/repos/db/index.go:677]: right, I will change them asap, to better reflect actual issue. Didn't want to block the release for now

File: adapters/repos/db/shard.go
Comment (jeroiraz) [inline adapters/repos/db/shard.go:1047]: haven't found where s.shut is set to true...

File: adapters/repos/db/shard.go
Comment (aliszka) [reply adapters/repos/db/shard.go:632]: i think this can be reverted. changed the value to verify it has the impact on pipeline failure

File: adapters/repos/db/shard.go
Comment (aliszka) [reply adapters/repos/db/shard.go:216]: it is written/read within `shutdownLock`

**Review by jeroiraz — APPROVED** [review]
LGTM only left the same comment over and over in different places to avoid hiding an error and kept consistency when the same function is called as in same places the original error is returned

File: adapters/repos/db/shard.go
Comment (aliszka) [reply adapters/repos/db/shard.go:1071]: As I understand method with context as an argument should respect context's timeout. Shutdown gets context, so if it reaches timeout it should fail and return error. Here timeout can be reached due to shard being in use for too long before even starting actual shutdown.

File: entities/locks/named_locks.go
Comment (aliszka) [reply entities/locks/named_locks.go:1]: oh, I didn't know about it. Will reuse it and remove my impl

File: adapters/repos/db/shard.go
Comment (aliszka) [reply adapters/repos/db/shard.go:1047]: inside `Shard::checkEligibleForShutdown`

**Review by moogacs — APPROVED** [review]

**Comment by moogacs** [comment]
LGTM, we can revisit in later on iteration to address comments to unblock the release while it's well tested

File: adapters/repos/db/migrator.go
Comment (moogacs) [reply adapters/repos/db/migrator.go:355]: resolve to merge

File: adapters/repos/db/index.go
Comment (moogacs) [reply adapters/repos/db/index.go:2179]: resolve to merge

File: adapters/repos/db/index.go
Comment (moogacs) [reply adapters/repos/db/index.go:2255]: resolve to merge

File: adapters/repos/db/shard.go
Comment (moogacs) [reply adapters/repos/db/shard.go:1071]: resolve to merge

File: adapters/repos/db/index.go
Comment (moogacs) [reply adapters/repos/db/index.go:2098]: resolve to merge

File: adapters/repos/db/index.go
Comment (moogacs) [reply adapters/repos/db/index.go:2120]: resolve to merge

File: adapters/repos/db/index.go
Comment (moogacs) [reply adapters/repos/db/index.go:2043]: resolve to merge

File: adapters/repos/db/index.go
Comment (moogacs) [reply adapters/repos/db/index.go:1872]: resolve to merge

File: adapters/repos/db/index.go
Comment (moogacs) [reply adapters/repos/db/index.go:1819]: resolve to merge

File: adapters/repos/db/index.go
Comment (moogacs) [reply adapters/repos/db/index.go:1690]: resolve to merge

File: adapters/repos/db/index.go
Comment (moogacs) [reply adapters/repos/db/index.go:1614]: resolve to merge

File: adapters/repos/db/shard.go
Comment (moogacs) [reply adapters/repos/db/shard.go:216]: resolve to merge

File: adapters/repos/db/index.go
Comment (moogacs) [reply adapters/repos/db/index.go:1191]: resolve to merge

File: adapters/repos/db/index.go
Comment (moogacs) [reply adapters/repos/db/index.go:1055]: resolve to merge

File: adapters/repos/db/shard.go
Comment (moogacs) [reply adapters/repos/db/shard.go:632]: resolve to merge

File: adapters/repos/db/index.go
Comment (moogacs) [reply adapters/repos/db/index.go:978]: resolve to merge

File: adapters/repos/db/index.go
Comment (moogacs) [reply adapters/repos/db/index.go:903]: resolve to merge

File: adapters/repos/db/index.go
Comment (moogacs) [reply adapters/repos/db/index.go:677]: resolve to merge

**Pull request merged**
